# -*- coding: utf-8 -*-
"""fetch_data

Automatically generated by Colab.

Original file is located at
    https://colab.research.google.com/drive/1KeUkwEAB5RDp0iZnr4TXQ4GQ0TF_Qqcv
"""

# Commented out IPython magic to ensure Python compatibility.
# %matplotlib inline
import os

import pandas as pd
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F

from Bio import SeqIO, Entrez
from Bio.Blast import NCBIWWW, NCBIXML
import requests

##Protein sequence retreival
!pip install biopython pandas requests

#Protein of interest: Aqua porin  https://www.uniprot.org/uniprotkb/A0A251P855/entry
uniprot_id = "A0A251P855"

# Fetch FASTA directly from UniProt
url = f"https://www.uniprot.org/uniprot/{uniprot_id}.fasta"
response = requests.get(url)
fasta_seq = response.text

# Run BLASTP against NCBI nr database
result_handle = NCBIWWW.qblast("blastp", "nr",  fasta_seq, hitlist_size=500)

Entrez.email = "sean011016@gmail.com"  # required by NCBI

blast_record = NCBIXML.read(result_handle)
print(len(blast_record))

results = []
for alignment in blast_record.alignments[:100]:  # top 5 hits
    acc = alignment.accession

    # Fetch full FASTA sequence from NCBI using accession
    handle = Entrez.efetch(db="protein", id=acc, rettype="fasta", retmode="text")
    record = SeqIO.read(handle, "fasta")
    handle.close()

    results.append([acc, str(record.seq)])

# Make DataFrame and save as CSV
seq_df = pd.DataFrame(results, columns=["Accession_ID", "Full_FASTA"])

print(seq_df.head())

seq_df.to_csv("blast_hits_raw.csv", index=False)